applications/asset_tracker_v2/src/modules/data_module.c

```diff
@@ -284,40 +300,173 @@ static void date_time_event_handler(const struct date_time_evt *evt)
+		if (list[i].ptr != NULL) {
+			k_free(list[i].ptr);
 
-			LOG_DBG("Pending data added: %p", ptr);
+			list[i].ptr = NULL;
+			list[i].len = 0;
+			list[i].type = 0;
 
-			return true;
 		}
 	}
 
-	LOG_WRN("Could not add pointer to pending list");
+	LOG_WRN("Data list cleared and freed");
+}
 
-	return false;
+static void data_list_add_failed(void *ptr, size_t len, enum data_type type)
+{
+retry:
+	for (size_t i = 0; i < ARRAY_SIZE(failed_data); i++) {
+		if (failed_data[i].ptr == NULL) {
+			failed_data[i].ptr = ptr;
+			failed_data[i].len = len;
+			failed_data[i].type = type;
+
+			LOG_DBG("Failed data added: %p", failed_data[i].ptr);
+			return;
+		}
+	}
+
+	LOG_WRN("Could not add data to failed data list, list is full");
+	LOG_WRN("Clearing failed data list and adding data");
+
+	data_list_clear_and_free(failed_data, ARRAY_SIZE(failed_data));
+	goto retry;
 }
 
-static bool pending_data_ack(void *ptr)
+static void data_list_add_pending(void *ptr, size_t len, enum data_type type)
 {
 	for (size_t i = 0; i < ARRAY_SIZE(pending_data); i++) {
-		if (pending_data[i] == ptr) {
-			k_free(ptr);
+		if (pending_data[i].ptr == NULL) {
+			pending_data[i].ptr = ptr;
+			pending_data[i].len = len;
+			pending_data[i].type = type;
 
-			LOG_DBG("Pending data ACKed: %p", pending_data[i]);
+			LOG_DBG("Pending data added: %p", pending_data[i].ptr);
+			return;
+		}
+	}
 
-			pending_data[i] = NULL;
+	LOG_ERR("Could not add data to pending data list, list is full");
+	SEND_ERROR(data, DATA_EVT_ERROR, -ENFILE);
+}
 
-			return true;
+static void data_resend(void)
+{
+	struct data_module_event *evt;
+
+	for (size_t i = 0; i < ARRAY_SIZE(failed_data); i++) {
+		if (failed_data[i].ptr != NULL) {
+			switch (failed_data[i].type) {
+			case DATA:
+				evt = new_data_module_event();
+				evt->type = DATA_EVT_DATA_SEND;
+				evt->data.buffer.buf = failed_data[i].ptr;
+				evt->data.buffer.len = failed_data[i].len;
+				LOG_WRN("Resending data: %.*s",
+					failed_data[i].len,
+					log_strdup(failed_data[i].ptr));
+				EVENT_SUBMIT(evt);
+				break;
+			case BATCH:
+				evt = new_data_module_event();
+				evt->type = DATA_EVT_DATA_SEND_BATCH;
+				evt->data.buffer.buf = failed_data[i].ptr;
+				evt->data.buffer.len = failed_data[i].len;
+				LOG_WRN("Resending batch data: %.*s",
+					failed_data[i].len,
+					log_strdup(failed_data[i].ptr));
+				EVENT_SUBMIT(evt);
+				break;
+			case CONFIG:
+				evt = new_data_module_event();
+				evt->type = DATA_EVT_CONFIG_SEND;
+				evt->data.buffer.buf = failed_data[i].ptr;
+				evt->data.buffer.len = failed_data[i].len;
+				LOG_WRN("Resending config data: %.*s",
+					failed_data[i].len,
+					log_strdup(failed_data[i].ptr));
+				EVENT_SUBMIT(evt);
+				break;
+			case UI:
+				evt = new_data_module_event();
+				evt->type = DATA_EVT_UI_DATA_SEND;
+				evt->data.buffer.buf = failed_data[i].ptr;
+				evt->data.buffer.len = failed_data[i].len;
+				LOG_WRN("Resending message data: %.*s",
+					failed_data[i].len,
```

Comment:
Looks like there's much in common for most cases here. So the event could be created outside the switch and only the type changed for each case. The default case can return, as it shouldn't add the data to pending list anyway